### ISSUE TITLE
7/9 keys active but elected is 9/9

### ISSUE THREAD

**Issue by gizemcakil** [harmony-one/harmony#2644]
When I stopped the node and stopped the AWS instance, my validator kept the slots associated with those instances. 

This seems like an easy arbitrage opportunity for validators because they don't 'actually' have to run nodes to keep their slots and earn.

my steps:

1. I have a validator with 9 keys, in 3 different t3.small instances
<img width="317" alt="image" src="https://user-images.githubusercontent.com/45907268/77997409-36757a00-72e4-11ea-811d-3b45f5335074.png">
2. I stopped the node by ctrl+z in tmux in the instance where I was running 2 keys in shard1
<img width="750" alt="Screen Shot 2020-03-30 at 9 48 25 PM" src="https://user-images.githubusercontent.com/45907268/77997502-60c73780-72e4-11ea-83bd-26e048f18856.png">
3. Waited for a few epochs, my validator still kept 9 elected slots

4. Killed the AWS instance that those 2 keys were running
<img width="817" alt="image" src="https://user-images.githubusercontent.com/45907268/77997578-848a7d80-72e4-11ea-8160-68001d84875f.png">
5. Validator still keeps 9 slots
<img width="290" alt="image" src="https://user-images.githubusercontent.com/45907268/77998472-0038fa00-72e6-11ea-949a-f5587aabad30.png">

Is it supposed to work this way?

**Comment by gizemcakil** [comment]
Adding what my CLI returns for validator information after the fact that i killed those instances. Please note that I didn't --remove-bls from the validator

  }
}
[ec2-user@ip-172-31-32-12 ~]$ ./hmy --node=https://api.s0.os.hmny.io/ blockchain validator information one1rcv3chw86tprvhpw4fjnpy2gnvqy4gp4fmhdd9
{
  "id": "0",
  "jsonrpc": "2.0",
  "result": {
    "current-epoch-performance": {
      "current-epoch-signing-percent": {
        "current-epoch-signed": 49,
        "current-epoch-signing-percentage": "0.777777777777777778",
        "current-epoch-to-sign": 63,
        "num-beacon-blocks-until-next-epoch": 32
      }
    },
    "currently-in-committee": true,
    "epos-status": "currently elected and signing enough blocks to be eligible for election next epoch",
    "lifetime": {
      "apr": "9.953755153666281778",
      "blocks": {
        "signed": 33040,
        "to-sign": 33976
      },
      "reward-accumulated": 4.851150212487437e+22
    },
    "metrics": {
      "by-shard": [
        {
          "bls-public-key": "47d8b40941cfff9daa245ed9d8a377790ac35ebe39b50a6f35d4abe2a99b1f6161cf8a28f36fabdb93f8709147339b08",
          "earning-account": "one1rcv3chw86tprvhpw4fjnpy2gnvqy4gp4fmhdd9",
          "effective-stake": "312224712500000000000000.000000000000000000",
          "group-percent": "0.034153550467891219",
          "overall-percent": "0.010929136149725190",
          "shard-id": 0
        },
        {
          "bls-public-key": "0f481d3d3109edfe112823906b1c37bfaef151621eca27d160590308b402aed2726cbe6b0166f1e51e825b646c998594",
          "earning-account": "one1rcv3chw86tprvhpw4fjnpy2gnvqy4gp4fmhdd9",
          "effective-stake": "312224712500000000000000.000000000000000000",
          "group-percent": "0.034153550467891219",
          "overall-percent": "0.010929136149725190",
          "shard-id": 0
        },
        {
          "bls-public-key": "806dd628feb8c9d9327405e9eb8f9a365f9b7367d9ad1450ee3de91dce3624e120e03b4426a7c80b0363d6dca65f6208",
          "earning-account": "one1rcv3chw86tprvhpw4fjnpy2gnvqy4gp4fmhdd9",
          "effective-stake": "312224712500000000000000.000000000000000000",
          "group-percent": "0.034153550467891219",
          "overall-percent": "0.010929136149725190",
          "shard-id": 0
        },
        {
          "bls-public-key": "ec5e4db27e44f436212f911e495f88face88b4c11a10952345155b11957d474cb7efdac5e463517967edbde832940590",
          "earning-account": "one1rcv3chw86tprvhpw4fjnpy2gnvqy4gp4fmhdd9",
          "effective-stake": "312224712500000000000000.000000000000000000",
          "group-percent": "0.034153550467891219",
          "overall-percent": "0.010929136149725190",
          "shard-id": 0
        },
        {
          "bls-public-key": "d57c71c055f7dec63e78cf17c50e114adaba91108b3ebc12d02d5e3b04608ab9239ba5ee5020599929fa85b7de78d119",
          "earning-account": "one1rcv3chw86tprvhpw4fjnpy2gnvqy4gp4fmhdd9",
          "effective-stake": "312224712500000000000000.000000000000000000",
          "group-percent": "0.044761679892059907",
          "overall-percent": "0.014323737565459170",
          "shard-id": 1
        },
        {
          "bls-public-key": "19662ac9ed6b29675fb89db112ce7a332e165caddc706c86bdf54a56c114829e08dcf1463fdb420e7f844b8dda278a8d",
          "earning-account": "one1rcv3chw86tprvhpw4fjnpy2gnvqy4gp4fmhdd9",
          "effective-stake": "312224712500000000000000.000000000000000000",
          "group-percent": "0.044761679892059907",
          "overall-percent": "0.014323737565459170",
          "shard-id": 1
        },
        {
          "bls-public-key": "0265f9764c0f724374351ea0b8e76ef07c3ea300ef1fb6ee92b5d65ddfbdcd0f39a776be05bfddfa509bf59976e75117",
          "earning-account": "one1rcv3chw86tprvhpw4fjnpy2gnvqy4gp4fmhdd9",
          "effective-stake": "312224712500000000000000.000000000000000000",
          "group-percent": "0.052193856083736658",
          "overall-percent": "0.016702033946795731",
          "shard-id": 3
        },
        {
          "bls-public-key": "70119d5a5c96866986c467527e6fbd3674158ec6c3de13e4b4e74fdc55635ae91dd389a1b3ebeaea260896cbaa060c0b",
          "earning-account": "one1rcv3chw86tprvhpw4fjnpy2gnvqy4gp4fmhdd9",
          "effective-stake": "312224712500000000000000.000000000000000000",
          "group-percent": "0.052193856083736658",
          "overall-percent": "0.016702033946795731",
          "shard-id": 3
        },
        {
          "bls-public-key": "7ac23db9d54fe085ff4ac2cc0f508121ca5eaebf213b1e99133acd175e563bff4c70a19146dd4908dda93d43653fae83",
          "earning-account": "one1rcv3chw86tprvhpw4fjnpy2gnvqy4gp4fmhdd9",
          "effective-stake": "312224712500000000000000.000000000000000000",
          "group-percent": "0.052193856083736658",
          "overall-percent": "0.016702033946795731",
          "shard-id": 3
        }
      ],
      "total-effective-stake": "2810022412500000000000000.000000000000000000"
    },
    "total-delegation": 6.333333e+24,
    "validator": {
      "address": "one1rcv3chw86tprvhpw4fjnpy2gnvqy4gp4fmhdd9",
      "bls-public-keys": [
        "d57c71c055f7dec63e78cf17c50e114adaba91108b3ebc12d02d5e3b04608ab9239ba5ee5020599929fa85b7de78d119",
        "19662ac9ed6b29675fb89db112ce7a332e165caddc706c86bdf54a56c114829e08dcf1463fdb420e7f844b8dda278a8d",
        "70119d5a5c96866986c467527e6fbd3674158ec6c3de13e4b4e74fdc55635ae91dd389a1b3ebeaea260896cbaa060c0b",
        "0265f9764c0f724374351ea0b8e76ef07c3ea300ef1fb6ee92b5d65ddfbdcd0f39a776be05bfddfa509bf59976e75117",
        "7ac23db9d54fe085ff4ac2cc0f508121ca5eaebf213b1e99133acd175e563bff4c70a19146dd4908dda93d43653fae83",
        "47d8b40941cfff9daa245ed9d8a377790ac35ebe39b50a6f35d4abe2a99b1f6161cf8a28f36fabdb93f8709147339b08",
        "ec5e4db27e44f436212f911e495f88face88b4c11a10952345155b11957d474cb7efdac5e463517967edbde832940590",
        "806dd628feb8c9d9327405e9eb8f9a365f9b7367d9ad1450ee3de91dce3624e120e03b4426a7c80b0363d6dca65f6208",
        "0f481d3d3109edfe112823906b1c37bfaef151621eca27d160590308b402aed2726cbe6b0166f1e51e825b646c998594"
      ],
      "creation-height": 926,
      "delegations": [
        {
          "amount": 6.333333e+24,
          "delegator-address": "one1rcv3chw86tprvhpw4fjnpy2gnvqy4gp4fmhdd9",
          "reward": 3.8269057847629725e+21,
          "undelegations": []
        }
      ],
      "details": "deets",
      "identity": "identoto",
      "last-epoch-in-committee": 127,
      "max-change-rate": "0.100000000000000000",
      "max-rate": "0.770000000000000000",
      "max-total-delegation": 1e+27,
      "min-self-delegation": 1.0001e+22,
      "name": "COVID-21, multiBLS (3 instances)",
      "rate": "0.170000000000000000",
      "security-contact": "CONTACT",
      "update-height": 926,
      "website": "gizem@harmony.one"
    }

**Comment by mindstyle85** [comment]
theyre still elected since their stake is high enough, but it does affect your signing:

"current-epoch-performance": {
      "current-epoch-signing-percent": {
        "current-epoch-signed": 252,
        "current-epoch-signing-percentage": "0.777777777777777778",
        "current-epoch-to-sign": 324,
        "num-beacon-blocks-until-next-epoch": 3

what strange is that if 3 out of 9 keys arent signing, means you should only be signing 2/3 of blocks (66,666%) and fall out of comittee i believe, and the fact that you sign with 77% is even more strange @rlan35 @fxfactorial ?

**Comment by mindstyle85** [comment]
actually this looks ok, youre signing with 7/9 keys, so thats 77,77777%:

"current-epoch-performance": {
"current-epoch-signing-percent": {
"current-epoch-signed": 252,
"current-epoch-signing-percentage": "0.777777777777777778",
"current-epoch-to-sign": 324,
"num-beacon-blocks-until-next-epoch": 3

and i think those 2 keys are still elected due to the high % of signing, just not sure if that should be the correct way since theyre inactive now? @rlan35 @gupadhyaya

**Comment by fxfactorial** [comment]
I think @gupadhyaya also answered on discord similarly, 
```
not a bug.... you are still signing 0.777777777777777778 (7 keys are signing 100% out of 9), hence your validator is not yet kicked out.
i recommend checking  ./hmy -n https://api.s0.os.hmny.io blockchain validator information one1rcv3chw86tprvhpw4fjnpy2gnvqy4gp4fmhdd9
```

**Comment by fxfactorial** [comment]
@gizemcakil Not a bug, this is supposed to be this way: 

because getting kicked out of committee selection is by validator one address, not per individual BLS key. A BLS key doesn't get kicked of election process, the validator does. 

So the validator's signing percentage was 

```
 "current-epoch-signing-percentage": "0.777777777777777778",
```

which is sufficient for the validator to be in the next EPoS election and hence whatever keys they want to use, sure, they can bring along with them. 

TL;DR we don't kick out BLS keys from committee election due to not signing. Instead we kick out one address of validator if their overal singin percentage < 66%.

**Comment by Dewansahil** [comment]
@fxfactorial the bls keys within the instance which was killed will still be sitting idle and blocking seats in that shard because the 'validator' is still 'elected'. This way a validator can block seats even without staking/signing?

**Comment by mindstyle85** [comment]
thats odd, when i tested with 3/4 keys active, the 3 were elected, and the one that wasnt signing was not elected, did this change since then?
im asking this because the 2 bls keys that arent signing here are still in the comittee and are using up seats for others that would actually sign
so unless im missing something it should be just 7/9 elected

**Comment by fxfactorial** [comment]
They paid for the slot, they can do what they want with it during the epoch. They aren't blocking seats, they won the auction and can do to do with it what they wish. So if they don't sign, they they are hurting themselves as well since that key isn't earning. Gizem's signing percent wasn't 100%, it was 77% so clearly there is some token reward not being earned.

I will try to extend the `hmy blockchain validator information <ADDR>` command so that you can see which key is earning and not, because atm its not clear which key is earning or not.

**Comment by mindstyle85** [comment]
while i can understand that, and this probably wont happen on a real network, but why let him stay in after 1 epoch of not signing with the key? its also bad for the network, not just for him?

**Comment by gupadhyaya** [comment]
@mindstyle85 whales cannot attack as long as they sign 2/3. If they don't they will get kicked out. Remember we only require 2/3.

**Comment by gizemcakil** [comment]
Who thinks this is a better design than kicking out those keys that are no longer signing?

Are there other trade-offs?

**Comment by fxfactorial** [comment]
By not signing then the validator is hurting themselves economically. Banned the key doesn't really mitigate it because they can just do it again with a new BLS key, you can always just generate a new key in the next epoch and adjust with edit-validator.

**Comment by mindstyle85** [comment]
> By not signing then the validator is hurting themselves economically. Banned the key doesn't really mitigate it because they can just do it again with a new BLS key, you can always just generate a new key in the next epoch and adjust with edit-validator.

ok this is a valid point, they can just edit it and add a new bls key, or activate their validator again

**Comment by mindstyle85** [comment]
@fxfactorial i was thinking about this.. not sure theres any better solution because of multiBLS here, so maybe at least make it back so that the non active bls keys arent elected (or stay elected), so that the owner of them has to at least notice that and edit validator to get them all back to active, seems to me its better than not making them at least do this (not necessarily for those that would intentionally leave keys inactive but those that might forget etc..)

**Comment by sophoah** [comment]
@mindstyle85 guess we would need some monitoring tools that will tell us if 1 or more of your BLS is not earning. And that would be possible with the latest update from @fxfactorial
